cli/src/commands/gerrit/send.rs

```diff
@@ -0,0 +1,353 @@
+    }
+
+    // the mapping is from old -> [new, is_dry_run]; the dry_run flag is used to
+    // disambiguate a later case when printing errors, so we know that if a
+    // commit was mapped to itself, it was because --dry-run was set, and not
+    // because e.g. it had an existing change id already
+    let mut old_to_new: IndexMap<Commit, (Commit, bool)> = IndexMap::new();
+    for commit_id in to_send.iter().map(|c| c.id()).rev() {
+        let original_commit = store.get_commit(commit_id).unwrap();
+        let description = original_commit.description().to_owned();
+        let footer = get_footer_lines(&description);
+
+        if !footer.is_empty() {
+            // first, figure out if there are multiple Change-Id fields; if so, then we
+            // error and continue
+            if footer.iter().filter(|entry| entry.0 == "Change-Id").count() > 1 {
+                writeln!(
+                    ui.warning(),
+                    "warning: multiple Change-Id footers in commit {}",
+                    short_commit_hash(original_commit.id()),
+                )?;
+                continue;
+            }
+
+            // now, look up the existing change id footer
+            let change_id = footer.iter().find(|entry| entry.0 == "Change-Id");
+            if let Some(FooterEntry(_, cid)) = change_id {
+                // map the old commit to itself
+                old_to_new.insert(original_commit.clone(), (original_commit.clone(), false));
+
+                // check the change-id format is correct in any case
+                if cid.len() != 41 || !cid.starts_with('I') {
+                    writeln!(
+                        ui.warning(),
+                        "warning: invalid Change-Id footer in commit {}",
+                        short_commit_hash(original_commit.id()),
+                    )?;
+                    continue;
+                }
+
+                // XXX (aseipp): should we rewrite these invalid Change-Ids? i
+                // don't think so, but I don't know what gerrit will do with
+                // them, and I realized my old signoff.sh script created invalid
+                // ones, so this is a helpful barrier.
+
+                continue; // fallthrough
+            }
+        }
+
+        if send.dry_run {
+            // mark the old commit as rewritten to itself, but only because it
+            // was a --dry-run, so we can give better error messages later
+            old_to_new.insert(original_commit.clone(), (original_commit.clone(), true));
+            continue;
+        }
+
+        // NOTE: Gerrit's change ID is not compatible with the alphabet used by
+        // jj, and the needed length of the change-id is different as well.
+        //
+        // for us, we convert to gerrit's format is the character 'I', followed
+        // by 40 characters of the truncated blake2 hash of the change id
+        let change_id = to_reverse_hex(&original_commit.change_id().hex()).unwrap();
```

Comment:
I ran into an issue which I think I can pinpoint as happening here. The issue was that after adding some extra changes or doing a rebase or something, gerrit wasn't happy with what I was pushing, so I closed the changes on gerrit, deleted the change id footers, and tried sending again.

But here, the gerrit change ID is calculated from the jj change ID, meaning that I got the same gerrit change IDs, preventing me from sending again. I'm not sure what deriving the gerrit ID from the jj ID buys us, but it does prevent me from easily doing what I mentioned.

Reply:
Yes, I've been meaning to fix this and ran into it myself. You need to `duplicate` then `abandon` to get a new Change ID. Frankly, I'll probably just take 4 random bytes and add them onto the hash, I just haven't gotten to it. But I can make this fix pretty quickly.

Reply:
This is now solved, by the way: the `Change-Id:` is now simply a hash of some random bytes, so you don't need to do something strange like `duplicate`/`abandon`.